src/Config/Config.php

```diff
@@ -0,0 +1,173 @@
+<?php
+
+/**
+ * Slim Framework (https://slimframework.com)
+ *
+ * @license https://github.com/slimphp/Slim-Console/blob/0.x/LICENSE.md (MIT License)
+ */
+
+declare(strict_types=1);
+
+namespace Slim\Console\Config;
+
+use ArrayAccess;
+
+use function array_key_exists;
+use function array_merge;
+use function ctype_lower;
+use function is_null;
+use function preg_replace;
+use function strtoupper;
+
+class Config implements ArrayAccess
+{
+    private const ENV_PREFIX = 'SLIM_CONSOLE_';
+
+    /**
+     * @var array<mixed>
+     */
+    private $default = [
+        'bootstrapDir'  => DIRECTORY_SEPARATOR . 'app',
+        'commandsDir'   => DIRECTORY_SEPARATOR . 'Application/Commands',
+        'indexDir'  => DIRECTORY_SEPARATOR . 'public',
+        'indexFile' => DIRECTORY_SEPARATOR . 'index.php',
+        'sourceDir' => DIRECTORY_SEPARATOR . 'src',
+    ];
+
+    /**
+     * @var array<mixed>
+     */
+    private $params = [];
+
+    /**
+     * Config constructor.
+     * @param array<mixed> $params
+     * @param string $configDir
+     */
+    public function __construct(array $params = [], string $configDir = '')
+    {
+        if ($configDir !== '') {
+            $this->set('rootDir', $configDir);
+        }
+        $this->setAll($params);
+    }
+
+    /**
+     * @return array<mixed>
+     */
+    public function all(): array
+    {
+        return $this->params;
+    }
+
+    /**
+     * @param array<mixed> $params
+     * @return void
+     */
+    public function setAll(array $params): void
+    {
+        $params = array_merge($this->default, $params);
+
+        foreach ($params as $key => $value) {
+            //Env variables take precedence
+            $this->params[$key] = $this->getEnvironmentVariableValue((string) $key) ?: $this->get('rootDir') . $value;
+        }
+    }
+
+    /**
+     * @param string $key
+     * @return mixed|null
+     */
+    public function get(string $key)
```

Comment:
I think that we should omit from using `ArrayAccess` and use named getters/setters:
- `getBootstrapDir(): string`
- `getCommandsDir(): ?string`
- `getIndexDir(): string`
- `getIndexFile(): string`
- `getSourceDir(): string`

Same for setters, here we return the `Config` object so we can chain `set` calls:
- `setBootstrapDir(string $dir): Config`
...

I think that this object should be immutable but extensible. So maybe an interface is necessary. Another point I would like to make is that in no instance should getters return null, as this will create downstream errors. These should all be validated, required parameters. The only parameter that could be null is `commandsDir`.

So instead of instantiating the configuration file with an array I would do something like:
```php
public function __construct(
    string $sourceDir,
    string $bootstrapDir,
    string $indexDir,
    string $indexFile,
    ?string $commandsDir = null,
)
```

Then you could have a static instantiator:
```php
public static function fromArray(array $params): Config
{
    // ... do array validation to ensure params contain required keys ...
   return new self(...);
}

public static function fromEnvironment(): Config
{
    // ... do array validation to ensure params contain required keys ...
   return new self(...);
}
```

Reply:
I agree.